pkg/gohai/network/network.go

```diff
@@ -6,29 +6,292 @@
 package network
 
-func getNetworkInfo() (networkInfo map[string]interface{}, err error) {
-	networkInfo = make(map[string]interface{})
+import (
+	"errors"
+	"fmt"
+	"net"
 
+	"github.com/DataDog/datadog-agent/pkg/gohai/utils"
+)
+
+// ErrAddressNotFound means no such address could be found
+var ErrAddressNotFound = errors.New("address not found")
+
+// Interface holds information about a specific interface
+type Interface struct {
+	// Name is the name of the interface
+	Name string `json:"name"`
+	// IPv4Network is the ipv4 address for the host
+	IPv4Network utils.Value[string] `json:"ipv4-network"`
+	// IPv6Network is the ipv6 address for the host
+	IPv6Network utils.Value[string] `json:"ipv6-network"`
+	// MacAddress is the mac address for the host
+	MacAddress utils.Value[string] `json:"macaddress"`
+	// IPv4 is the list of IPv4 addresses for the interface
+	IPv4 []string `json:"ipv4"`
+	// IPv4 is the list of IPv6 addresses for the interface
+	IPv6 []string `json:"ipv6"`
+}
+
+// Info holds network metadata about the host
+type Info struct {
+	// Interfaces is the list of interfaces which are up
+	Interfaces []Interface `json:"interfaces"`
+	// MacAddress is a mac address of the host
+	MacAddress string `json:"macaddress"`
+	// IPAddress is an IPv4 address of the host
+	IPAddress string `json:"ipaddress"`
+	// IPAddressV6 is an IPv6 address of the host
+	IPAddressV6 utils.Value[string] `json:"ipaddressv6"`
+}
+
+// CollectInfo collects the network information.
+func CollectInfo() (*Info, error) {
+	info := &Info{}
+	err := fillNetworkInfo(info)
+	if err != nil {
+		return info, err
+	}
+
+	interfaces, err := getMultiNetworkInfo()
+	if err == nil && len(interfaces) > 0 {
+		info.Interfaces = interfaces
+	}
+	return info, err
+}
+
+func ifacesToJSON(ifaces []Interface) (interface{}, []string) {
+	ret := make([]interface{}, len(ifaces))
+	warnings := []string{}
+
+	for idx, iface := range ifaces {
+		ifaceJSON := map[string]interface{}{
+			"name": iface.Name,
+			"ipv4": iface.IPv4,
+			"ipv6": iface.IPv6,
+		}
+
+		values := map[string]utils.Value[string]{
+			"ipv4-network": iface.IPv4Network,
+			"ipv6-network": iface.IPv6Network,
+			"macaddress":   iface.MacAddress,
+		}
+		for key, value := range values {
+			if val, err := value.Value(); err == nil {
+				ifaceJSON[key] = val
+			} else {
+				warnings = append(warnings, fmt.Sprintf("%s: %s: %v", iface.Name, key, err))
+			}
+		}
+
+		ret[idx] = ifaceJSON
+	}
+
+	return ret, warnings
+}
+
+// AsJSON returns an interface which can be marshalled to a JSON and contains the value of non-errored fields.
+func (netInfo *Info) AsJSON() (interface{}, []string, error) {
+	interfaces, warnings := ifacesToJSON(netInfo.Interfaces)
+	ret := map[string]interface{}{
+		"macaddress": netInfo.MacAddress,
+		"ipaddress":  netInfo.IPAddress,
+		"interfaces": interfaces,
+	}
+
+	if ipv6, err := netInfo.IPAddressV6.Value(); err == nil {
+		ret["ipaddressv6"] = ipv6
```

Comment:
Could this be handled by the `utils.Value` type ? This way we could loop over the `utils.Value` for warnings and then use `json.Marshal`.

Reply:
As explained in this [comment](https://github.com/DataDog/datadog-agent/pull/18682#discussion_r1290139036), values shouldn't be included if they are an error, and you can't dynamically say whether a type should be included in the output JSON, the closest thing is the `omitempty` field tag but that doesn't work if the value is not the default.
So redefining how the whole thing is marshaled is the only way